### PR TITLE
[TEST] Add test for Aggregations#fromXContent

### PR DESCRIPTION
**Pull request by javanna**
AggregationsTests#testFromXContent verifies that parsing of aggregations works by combining multiple aggs at the same level, and also adding sub-aggregations to multi bucket and single bucket aggs, up to a maximum depth of 5.

### CODE REVIEW

File: core/src/test/java/org/elasticsearch/search/aggregations/AggregationsTests.java [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+import org.junit.After;
+import org.junit.Before;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+import static java.util.Collections.singletonMap;
+
+public class AggregationsTests extends ESTestCase {
```
Comment (tlrx) [inline core/src/test/java/org/elasticsearch/search/aggregations/AggregationsTests.java]: Can you add some doc about the purpose of this test?

File: core/src/test/java/org/elasticsearch/search/aggregations/AggregationsTests.java [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        aggsTests.add(new InternalBucketMetricValueTests());
+        aggsTests.add(new InternalStatsTests());
+        aggsTests.add(new InternalStatsBucketTests());
+        aggsTests.add(new InternalExtendedStatsTests());
+        aggsTests.add(new InternalExtendedStatsBucketTests());
+        aggsTests.add(new InternalGeoBoundsTests());
+        aggsTests.add(new InternalGeoCentroidTests());
+        aggsTests.add(new InternalHistogramTests());
+        aggsTests.add(new InternalDateHistogramTests());
+        return aggsTests;
```
Comment (tlrx) [inline core/src/test/java/org/elasticsearch/search/aggregations/AggregationsTests.java]: nit: can it be an unmodifiable list?

File: core/src/test/java/org/elasticsearch/search/aggregations/AggregationsTests.java [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+            BytesReference parsedBytes = XContentHelper.toXContent(parsedAggregations, xContentType, randomBoolean());
+            ElasticsearchAssertions.assertToXContentEquivalent(originalBytes, parsedBytes, xContentType);
+        }
+    }
+
+    private static InternalAggregations createTestInstance() {
+        return createTestInstance(0, 5);
+    }
+
+    private static InternalAggregations createTestInstance(int currentDepth, int maxDepth) {
```
Comment (tlrx) [inline core/src/test/java/org/elasticsearch/search/aggregations/AggregationsTests.java]: maxDepth can be final

File: core/src/test/java/org/elasticsearch/search/aggregations/InternalMultiBucketAggregationTestCase.java [hunk 4]
```diff
@@ -21,36 +21,42 @@
...
+    @Override
+    public void setUp() throws Exception {
+        super.setUp();
+        if (randomBoolean()) {
+            subAggregationsSupplier = () -> InternalAggregations.EMPTY;
+        } else {
+            subAggregationsSupplier = () -> {
+                final int numAggregations = randomIntBetween(1, 3);
+                List<InternalAggregation> aggs = new ArrayList<>();
+                for (int i = 0; i <numAggregations; i++) {
```
Comment (tlrx) [inline core/src/test/java/org/elasticsearch/search/aggregations/InternalMultiBucketAggregationTestCase.java]: space after  ̀<`

File: core/src/test/java/org/elasticsearch/search/aggregations/metrics/cardinality/InternalCardinalityTests.java [hunk 5]
```diff
@@ -38,13 +37,21 @@
     private static List<HyperLogLogPlusPlus> algos;
     private static int p;
 
-    @Before
-    public void setup() {
+    @After //we force @After to have it run before ESTestCase#after otherwise it fails
```
Comment (tlrx) [inline core/src/test/java/org/elasticsearch/search/aggregations/metrics/cardinality/InternalCardinalityTests.java]: I don't understand why it is @After and not @Before anymore... This is going to break the test.

**Review by tlrx — CHANGES_REQUESTED** [review]
This looks very nice and elegant, I like it a lot. I requested a change about a @After annotation but once it is fixed it will be good.

File: core/src/test/java/org/elasticsearch/search/aggregations/metrics/cardinality/InternalCardinalityTests.java (hunk 5 above)
Comment (javanna) [inline core/src/test/java/org/elasticsearch/search/aggregations/metrics/cardinality/InternalCardinalityTests.java]: it does not make any sense. let me dig why tests run though... I thought after needed to be forced on teardDown, let me verify that!

File: core/src/test/java/org/elasticsearch/search/aggregations/metrics/cardinality/InternalCardinalityTests.java [hunk 6]
```diff
@@ -38,13 +37,21 @@
...
+    @Override
+    public void setUp() {
```
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/search/aggregations/metrics/cardinality/InternalCardinalityTests.java]: Does this need to call super.setUp() as well as it overrides LuceneTestCase#setUp()?

File: core/src/test/java/org/elasticsearch/search/aggregations/metrics/cardinality/InternalCardinalityTests.java [hunk 7]
```diff
@@ -38,13 +37,21 @@
...
         algos = new ArrayList<>();
         p = randomIntBetween(HyperLogLogPlusPlus.MIN_PRECISION, HyperLogLogPlusPlus.MAX_PRECISION);
     }
 
     @Override
+    public void tearDown() {
```
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/search/aggregations/metrics/cardinality/InternalCardinalityTests.java]: Does this need to call super.tearDown() as well as it overrides LuceneTestCase#tearDown()?

**Review by cbuescher — COMMENTED** [review]
I left two comment regarding setUp/tearDown. I generally find that part a bit hard to follow, but I see why it is done this way.

File: core/src/test/java/org/elasticsearch/search/aggregations/metrics/cardinality/InternalCardinalityTests.java (hunk 6 above)
Comment (javanna) [inline core/src/test/java/org/elasticsearch/search/aggregations/metrics/cardinality/InternalCardinalityTests.java]: yes it does. I have messed these two up for some reason. Probably ran only AggregationsTests and not the cardinality tests...fixing

File: core/src/test/java/org/elasticsearch/search/aggregations/metrics/cardinality/InternalCardinalityTests.java (hunk 7 above)
Comment (javanna) [inline core/src/test/java/org/elasticsearch/search/aggregations/metrics/cardinality/InternalCardinalityTests.java]: yep

**Comment by javanna** [comment]
> I generally find that part a bit hard to follow, but I see why it is done this way.

I agree, and that is because there are multiple ways to do the same thing, but each one works slightly differently :)  I wanted to have one single way to set things up in the base test class. THen I hooked into the lucene setup and teardown methods. That also makes sure that we don't forget to call super which is nice. Yet given that the annotations are in the base class LuceneTestCase, the after in ESTestCase runs first, and that is why in one case we have to force the After annotation, so that one is called first.

**Comment by javanna** [comment]
@tlrx @cbuescher I have addressed your comments, can you have another look please?

**Review by tlrx — APPROVED** [review]

**Review by cbuescher — APPROVED** [review]
